Tests/Tests/STPRedirectContextTest.m

```diff
@@ -77,6 +79,125 @@ - (void)testInitWithConsumedSourceReturnsNil {
+    }];
+
+    // Make sure the initWithSource: method pulled out the right values from the Source
+    XCTAssertNil(sut.nativeRedirectUrl);
+    XCTAssertEqualObjects(sut.redirectUrl, source.redirect.url);
+    XCTAssertEqualObjects(sut.returnUrl, source.redirect.returnURL);
+
+    // and make sure the completion calls the completion block above
+    sut.completion(fakeError);
+    [self waitForExpectationsWithTimeout:0 handler:nil];
+}
+
+- (void)testInitWithSourceWithNativeURL {
+    STPSource *source = [STPFixtures alipaySourceWithNativeUrl];
+    XCTestExpectation *expect = [self expectationWithDescription:@"completion"];
+    NSURL *nativeURL = [NSURL URLWithString:source.details[@"native_url"]];
+    NSError *fakeError = [NSError new];
+
+    STPRedirectContext *sut = [[STPRedirectContext alloc] initWithSource:source completion:^(NSString * _Nonnull sourceID, NSString * _Nullable clientSecret, NSError * _Nullable error) {
+        XCTAssertEqualObjects(source.stripeID, sourceID);
+        XCTAssertEqualObjects(source.clientSecret, clientSecret);
+        XCTAssertEqual(error, fakeError, @"Should be the same NSError object passed to completion() below");
+        [expect fulfill];
+    }];
+
+    // Make sure the initWithSource: method pulled out the right values from the Source
+    XCTAssertEqualObjects(sut.nativeRedirectUrl, nativeURL);
+    XCTAssertEqualObjects(sut.redirectUrl, source.redirect.url);
+    XCTAssertEqualObjects(sut.returnUrl, source.redirect.returnURL);
+
+    // and make sure the completion calls the completion block above
+    sut.completion(fakeError);
+    [self waitForExpectationsWithTimeout:0 handler:nil];
+}
+
+- (void)testInitWithPaymentIntent {
+    STPPaymentIntent *paymentIntent = [STPFixtures paymentIntent];
+    XCTestExpectation *expect = [self expectationWithDescription:@"completion"];
+    NSError *fakeError = [NSError new];
+    NSString *returnUrlString = @"payments-example://stripe-redirect";
+
+    STPRedirectContext *sut = [[STPRedirectContext alloc] initWithPaymentIntent:paymentIntent returnUrl:returnUrlString completion:^(NSString * _Nonnull clientSecret, NSError * _Nullable error) {
+        XCTAssertEqualObjects(paymentIntent.clientSecret, clientSecret);
+        XCTAssertEqual(error, fakeError, @"Should be the same NSError object passed to completion() below");
+        [expect fulfill];
+    }];
+
+    // Make sure the initWithPaymentIntent: method pulled out the right values from the PaymentIntent
+    XCTAssertNil(sut.nativeRedirectUrl);
+    XCTAssertEqualObjects(sut.redirectUrl.absoluteString,
+                          @"https://hooks.stripe.com/redirect/authenticate/src_1Cl1AeIl4IdHmuTb1L7x083A?client_secret=src_client_secret_DBNwUe9qHteqJ8qQBwNWiigk");
+    XCTAssertEqualObjects(sut.returnUrl.absoluteString, returnUrlString);
+
+    // and make sure the completion calls the completion block above
+    sut.completion(fakeError);
+    [self waitForExpectationsWithTimeout:0 handler:nil];
```

Comment:
hmm, since these are using a timeout value of 0, why not just use a `__block BOOL completionCalled` to avoid any async issues of `waitFor...`?

Reply:
I blanked on what the right pattern for that was. Thanks!